internal/auth/oidc/repository_auth_method_update.go

```diff
@@ -17,32 +31,559 @@ import "context"
+			txRepo := &Repository{
+				reader: reader,
+				writer: w,
+				kms:    r.kms,
+				// intentionally not setting the defaultLimit, so we'll get all
+				// the account ids without a limit
+			}
+			updatedAm, err = txRepo.lookupAuthMethod(ctx, updatedAm.PublicId)
+			if err != nil {
+				return errors.Wrap(err, op, errors.WithMsg("unable to lookup auth method after update"))
+			}
+			return nil
+		},
+	)
+	if err != nil {
+		return nil, db.NoRowsAffected, errors.Wrap(err, op)
+	}
+
+	providerCache().delete(ctx, updatedAm.PublicId)
+
+	return updatedAm, rowsUpdated, nil
+}
+
+// voName represents the names of auth method value objects
+type voName string
+
+const (
+	SigningAlgVO  voName = "SigningAlgs"
+	CallbackUrlVO voName = "CallbackUrls"
+	CertificateVO voName = "Certificates"
+	AudClaimVO    voName = "AudClaims"
+)
+
+// validVoName decides if the name is valid
+func validVoName(name voName) bool {
+	switch name {
+	case SigningAlgVO, CallbackUrlVO, CertificateVO, AudClaimVO:
+		return true
+	default:
+		return false
+	}
+}
+
+// factoryFunc defines a func type for value object factories
+type factoryFunc func(publicId string, i interface{}) (interface{}, error)
+
+// supportedFactories are the currently supported factoryFunc for value objects
+var supportedFactories = map[voName]factoryFunc{
+	SigningAlgVO: func(publicId string, i interface{}) (interface{}, error) {
+		str := fmt.Sprintf("%s", i)
+		return NewSigningAlg(publicId, Alg(str))
+	},
+	CertificateVO: func(publicId string, i interface{}) (interface{}, error) {
+		str := fmt.Sprintf("%s", i)
+		return NewCertificate(publicId, str)
+	},
+	AudClaimVO: func(publicId string, i interface{}) (interface{}, error) {
+		str := fmt.Sprintf("%s", i)
+		return NewAudClaim(publicId, str)
+	},
+	CallbackUrlVO: func(publicId string, i interface{}) (interface{}, error) {
+		u, err := url.Parse(fmt.Sprintf("%s", i))
+		if err != nil {
+			return nil, err
+		}
+		return NewCallbackUrl(publicId, u)
+	},
+}
+
+// valueObjectChanges takes the new and old list of VOs (value objects) and
+// using the dbMasks/nullFields it will return lists of VOs where need to be
+// added and deleted in order to reconcile auth method's value objects.
+func valueObjectChanges(
+	publicId string,
+	valueObjectName voName,
+	newVOs,
+	oldVOs,
+	dbMask,
+	nullFields []string,
+) (add []interface{}, del []interface{}, e error) {
+	const op = "valueObjectChanges"
+	if publicId == "" {
+		return nil, nil, errors.New(errors.InvalidParameter, op, "missing public id")
+	}
+	if !validVoName(valueObjectName) {
+		return nil, nil, errors.New(errors.InvalidParameter, op, fmt.Sprintf("invalid value object name: %s", valueObjectName))
+	}
+	if !strutil.StrListContains(dbMask, string(valueObjectName)) && !strutil.StrListContains(nullFields, string(valueObjectName)) {
+		return nil, nil, nil
+	}
+	if len(strutil.RemoveDuplicates(newVOs, false)) != len(newVOs) {
```

Comment:
Some value objects don't care about removing case insensitive duplicates while others do.  Should this function allow the caller to specify?

Reply:
All value objects (Callbacks, Certs, Algs, and Auds must be unique or the database will reject them based on the primary keys of their storage tables.  That's this function raises an error if they are not.